src/core/thread/network_diagnostic.cpp

```diff
@@ -224,15 +224,21 @@ otError NetworkDiagnostic::AppendIp6AddressList(Message &aMessage)
 otError NetworkDiagnostic::AppendChildTable(Message &aMessage)
 {
     otError         error   = OT_ERROR_NONE;
-    uint8_t         count   = 0;
+    uint16_t        count   = 0;
     uint8_t         timeout = 0;
     ChildTableTlv   tlv;
     ChildTableEntry entry;
 
     tlv.Init();
 
     count = Get<ChildTable>().GetNumChildren(ChildTable::kInStateValid);
-    tlv.SetLength(count * sizeof(ChildTableEntry));
+
+    if (count > (CHAR_BIT * sizeof(uint8_t) / sizeof(ChildTableEntry)))
```

Comment:
@jwhui I believe this part of code produces incorrect TLV that affects Router 5.7.3 certification test case (TLV length different than payload size).

What was the purpose of this if statement? During review I thought the idea was to limit the TLV length to 255, but this code actually does not do it.

Reply:
@LuDuda, thanks for raising this issue. I'm not sure what's going on here :)

Would you like to submit a PR with the fix?

Reply:
I see a couple of options:

a) make sure that TLV length never exceeds 255 (limits child table TLV to ~85 entries)
b) introduce extended TLV when needed (limits child table TLV to network MTU and so ~400 entries but hard to tell what is the exact limit)
c) implement a) or b) and file Thread JIRA on how to handle situation where there is e.g. 500 children attached (spec requires to append information about all the children).

What do you think about option c)?

Reply:
Thanks @LuDuda for bringing this up.  I'd vote for doing (a) as first step, followed by (c). 

Thinking about this, we may face other similar situations which we need to address. For example running out of the spinel NCP buffer size when encoding the child table for `SPINEL_THREAD_CHILD_TABLE` property - maybe need to think of a solution to send partial table?.

Reply:
Agree with option (c) - implement (a) and file Thread Group JIRA for additional clarify.